File: samples/mapzen-sample/src/main/java/mapzen/com/sdksampleapp/activities/MainActivity.java
package mapzen.com.sdksampleapp.activities;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;
import mapzen.com.sdksampleapp.R;
import mapzen.com.sdksampleapp.models.Sample;
import mapzen.com.sdksampleapp.models.SampleMap;

/**
 * Entry point for the sample app. Displays bottom navigation bar with top scroll view for
 * interaction with different SDK use cases.
 */
public class MainActivity extends AppCompatActivity implements
    View.OnClickListener {

  Sample current = null;

  @Override protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
    setupBottomNavigation();
    navbarItemSelected(R.id.navigation_map);
  }

  private void setupBottomNavigation() {
    BottomNavigationView nav = findViewById(R.id.navigation);
    nav.setOnNavigationItemSelectedListener(
        new BottomNavigationView.OnNavigationItemSelectedListener() {
      @Override public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        navbarItemSelected(item.getItemId());
        return true;
      }
    });
  }

  private void configureScrollView(Sample[] samples) {
    LinearLayout layout = findViewById(R.id.scrollContent);
    layout.removeAllViews();
    if (samples == null) {
      return;
    }
    for (Sample sample : samples) {
      LayoutInflater inflater = LayoutInflater.from(this);
      TextView textView = (TextView) inflater.inflate(R.layout.text_row, null);
      textView.setText(sample.title);
      textView.setOnClickListener(this);
      textView.setTag(sample);
      LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(WRAP_CONTENT,
          WRAP_CONTENT);
      if (sample == samples[samples.length - 1]) {
        int rightMargin = getResources().getDimensionPixelSize(R.dimen.padding_large);
        layoutParams.setMargins(0, 0, rightMargin, 0);
      }
      layout.addView(textView, layoutParams);
    }
  }

  @Override public void onClick(View view) {
    Sample sample = (Sample) view.getTag();
    current.takedown(null, null, null);
    current = sample;
    current.setup(null, null, null);
  }

  private void navbarItemSelected(int itemId) {
    Sample[] samples = SampleMap.NAV_ID_TO_SECTIONS.get(itemId);
    configureScrollView(samples);
    if (current != null) {
      current.takedown(null, null, null); //TODO
      current = null;
    }
    if (samples != null && samples.length > 0) {
      current = samples[0];
    }
    if (current != null) {
      current.setup(null, null, null); //TODO
    }
  }
}

File: samples/mapzen-sample/src/main/java/mapzen/com/sdksampleapp/activities/MainActivity.kt
package mapzen.com.sdksampleapp.activities

import android.content.Intent
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.design.widget.BottomNavigationView
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.HorizontalScrollView
import android.widget.LinearLayout
import android.widget.TextView
import kotterknife.bindView
import mapzen.com.sdksampleapp.R
import mapzen.com.sdksampleapp.controllers.MainController
import mapzen.com.sdksampleapp.models.Sample
import mapzen.com.sdksampleapp.presenters.MainPresenter
import javax.inject.Inject

/**
 * Entry point for the sample app. Displays bottom navigation bar with top scroll view for
 * interaction with different SDK use cases.
 */
class MainActivity : BaseActivity(), MainController {

  val navigationView: BottomNavigationView by bindView(R.id.navigation)
  val scrollContent: LinearLayout by bindView(R.id.scrollContent)
  val scrollView: HorizontalScrollView by bindView(R.id.scrollView)

  @Inject lateinit var presenter: MainPresenter

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    mainApplication.appComponent.inject(this)
    presenter.controller = this
    presenter.onCreate()
    PreferenceManager.setDefaultValues(this, R.xml.preferences, false)
  }

  override fun onDestroy() {
    presenter.onDestroy()
    super.onDestroy()
  }

  override fun onCreateOptionsMenu(menu: Menu?): Boolean {
    menuInflater.inflate(R.menu.activity_main, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem?): Boolean {
    presenter.onOptionsItemSelected(item?.itemId)
    return true
  }

  override fun selectSampleView(sample: Sample) {
    (0..scrollContent.childCount)
      .filter { it -> scrollContent.getChildAt(it) is TextView }
      .map { scrollContent.getChildAt(it) }
      .forEach { it.isSelected = (it.tag == sample) }
  }

  override fun setupNavigationItemSelectedListener() {
    navigationView?.setOnNavigationItemSelectedListener { item ->
      presenter.onNavBarItemSelected(item.itemId)
      true
    }
  }

  override fun cleanupNavigationItemSelectedListener() {
    navigationView?.setOnNavigationItemSelectedListener(null)
  }

  override fun clearScrollViewSamples() {
    scrollContent?.removeAllViews()
  }

  override fun cleanupSampleFragment() {
    val fragment = supportFragmentManager.findFragmentById(R.id.fragment)
    fragment?.let { supportFragmentManager.beginTransaction().remove(it).commit() }
  }

  override fun setupSampleFragment(sample: Sample) {
    val fragment = sample.fragmentClass.java.newInstance()
    supportFragmentManager.beginTransaction().replace(R.id.fragment, fragment).commit()
  }

  override fun setScrollViewSamples(samples: Array<Sample>?) {
    if (samples == null) { return }
    for (sample in samples) {
      val inflater = LayoutInflater.from(this)
      val textView = inflater.inflate(R.layout.text_row, null) as TextView
      textView.text = presenter?.getTitleText(sample)
      textView.setOnClickListener { view ->
        scrollView.smoothScrollTo(view.x.toInt(), 0)
        val sample = view.tag as Sample
        presenter?.onSampleSelected(sample)
      }
      textView.tag = presenter?.getTag(sample)
      val layoutParams = LinearLayout.LayoutParams(WRAP_CONTENT, WRAP_CONTENT)
      if (sample != samples[samples.size - 1]) {
        val rightMargin = resources.getDimensionPixelSize(R.dimen.padding)
        layoutParams.setMargins(0, 0, rightMargin, 0)
      }
      scrollContent?.addView(textView, layoutParams)
    }
  }

  override fun cleanupScrollItemClickListeners() {
    (0..scrollContent.childCount)
      .filter { it -> scrollContent.getChildAt(it) is TextView }
      .map { scrollContent.getChildAt(it) as TextView }
      .forEach { it.setOnClickListener(null) }
  }

  override fun openSettings() {
    val intent = Intent(this, SettingsActivity::class.java)
    startActivity(intent)
  }
}
